### ISSUE TITLE
ArmedAssault3 no response

### ISSUE THREAD

**Issue by senshi-x** [Austinb/GameQ#233]
I want to set up GQv3 to show info on our ArmA3 server. I use the v3.0.0alpha2 release.

However, I don't get a response from the A3 server at all. And I have no idea why.
The same server is listed fine on [GameTracker](http://www.gametracker.com/server_info/144.76.140.250:2302/).

See [here](https://www.wargamer-project.de/gameq/examples/index_broken.php) for the current example.

I also don't get any kind of special information despite using the debug option.

I also tried setting up the query on our TS3 server, which then worked just fine (see the commented section in the servers array).

It is absolutely possible that this is a mistake on my end, as I know next to nothing about PHP and just fiddled through the examples to get it to work without PHP errors...Maybe one of you can help me?

``` php
<?php
ini_set('display_errors',1);
ini_set('display_startup_errors',1);
error_reporting(-1);
// Include the main class file
require ('../src/GameQ/Autoloader.php');
require ('../src/GameQ/GameQ.php');
// Define your servers,
// see list.php for all supported games and identifiers.
$servers = array(
    array(
        'id' => 'ArmA3',
        'type' => 'armedassault3',
        'host' => '144.76.140.250:2302',
        //'host' => '37.187.170.178:2302'
    ),  
    /*array(
        'id' => 'TS3',
        'type' => 'teamspeak3',
        //'host' => '144.76.140.250:2302',
        'host' => '85.214.55.198:9988',
        'options' => [
            'query_port' => 10011,
        ],
    )*/
);

// Call the class, and add your servers.
$gq = new \GameQ\GameQ();
$gq->addServers($servers);

// You can optionally specify some settings
$gq->setOption('timeout', 10); // Seconds
$gq->setOption('debug', true);

// You can optionally specify some output filters,
// these will be applied to the results obtained.

// Send requests, and parse the data
$results = $gq->process();


// Some functions to print the results
function print_results($results) {

    print_r($results);
    foreach ($results as $id => $data) {

        printf("<h2>%s</h2>\n", $id);
        print_table($data);
    }

}

function print_table($data) {

    $gqs = array('gq_online', 'gq_address', 'gq_port', 'gq_prot', 'gq_type');


    if (!$data['gq_online']) {
        printf("<p>The server did not respond within the specified time.</p>\n");
        //return;
    }

    print("<table><thead><tr><td>Variable</td><td>Value</td></tr></thead><tbody>\n");

    foreach ($data as $key => $val) {

        if (is_array($val)) continue;

        $cls = empty($cls) ? ' class="uneven"' : '';

        if (substr($key, 0, 3) == 'gq_') {
            $kcls = (in_array($key, $gqs)) ? 'always' : 'normalise';
            $key = sprintf("<span class=\"key-%s\">%s</span>", $kcls, $key);
        }

        printf("<tr%s><td>%s</td><td>%s</td></tr>\n", $cls, $key, $val);
    }

    print("</tbody></table>\n");

}

?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
    <head>
        <title>GameQ - Example script</title>
        <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
        <style type="text/css">
            * {
                font-size: 9pt;
                font-family: Verdana, sans-serif;
            }
            h1 {
                font-size: 12pt;
            }
            h2 {
                margin-top:2em;
                font-size: 10pt;
            }
            table {
                border: 1px solid #000;
                background-color: #DDD;
                border-spacing:1px 1px;
            }
            table thead {
                font-weight: bold;
                background-color: #CCC;
            }
            table tr.uneven td {
                background-color:#FFF;
            }
            table td {
                padding: 5px 8px;
            }
            table tbody {
                background-color: #F9F9F9;
            }
            .note {
                color: #333;
                font-style:italic;
            }
            .key-always {
                color:red;
                font-weight:bold;
            }
            .key-normalise {
                color:red;
            }
        </style>
    </head>
    <body>
    <h1>GameQ - Example script</h1>
    <div class="note">
    Players are never displayed in this example. <br/>
    <span class="key-always">Bold, red</span> variables are always set by gameq.
    Additionally, the normal <span class="key-normalise">red</span> variables are always set when the normalise filter is enabled.<br/>
    gq_online will always contain a boolean indicating if the server responded to the request.<br/>
    </div>
<?php
    print_results($results);
?>
    </body>
</html>
```

**Comment by Austinb** [comment]
I am able to see the server output using the following code:

``` php
$gq = \GameQ\GameQ::factory();
$gq->addServer([
    'type' => 'armedassault3',
    'host' => '144.76.140.250:2302',
]);
$results = $gq->process();
print_r($results);
```

Also note that you only need to include the Autoloader.php file if you are not using Composer.  The extra include for GameQ.php is redundant. 

Edit:  Using the following worked and show me both.

``` php
$servers = array(
    array(
        'id' => 'ArmA3',
        'type' => 'armedassault3',
        'host' => '144.76.140.250:2302',
    ),
    array(
        'id' => 'TS3',
        'type' => 'teamspeak3',
        'host' => '85.214.55.198:9988',
        'options' => [
            'query_port' => 10011,
        ],
    )
);

$gq->addServers($servers);
```

**Comment by senshi-x** [comment]
Thanks for the reply.

But it's odd that it just works for you.

I copied your code verbatim and it isn't working for me. When comparing your version to mine, the only difference I see is your init call anyway:

> $gq = \GameQ\GameQ::factory();

instead of my

> $gq = new \GameQ\GameQ();

I do get the TS3 output (as I did before), but nothing for the Arma3 server.

EDIT: PHP version is 5.5.18.

EDIT2: Update. I tested this now on a local XAMPP server, and there it works just fine. So it looks like it's an issue with where it's hosted (or possibly the PHP version, as I have 5.5.19 locally, but that seems unlikely).

**Comment by Austinb** [comment]
There is no real difference in the call to the factory vs the new class() call.  

Check your php version on the remote server using phpinfo(); or calling `php -v` in shell if you have access.  It is also possible the host is blocking the ports in one or both directions.  Many shared hosting providers do this to prevent abuse and/or the use of bots.

**Comment by Austinb** [comment]
Since this appears to be a hosting/open ports issue closing.

**Comment by senshi-x** [comment]
Just to "solve" this:
It indeed was a firewall issue. I will put the relevant communication with the server hosting company (CTO got involved, even :) ) here, if someone else stumbles on this issue.

tl;dr version: It seems GameQ can be caught in very restrictive firewalls due to some imperfect UDP state handling.

Reply on initial query by CTO:

```
We are very restrictive on unsolicited UDP at our firewalls... We use "PINHOLE"
type methodology in that if one of our server sends a UDP packet, then the
replies will be allowed back in using 'state' based on originating and symmetrical
source/destination IP and ports otherwise they are dropped on the floor... This was
done back in 2009 to stop a whole class of attacks and floods from new types of
scanning worms and zombies using UDP for discovery/communication...

I would need to know more about its UDP communication flow before I can proceed
further...
```

After the requested clarification was sent:

```
I have punched a 'stateful' hole in our PINHOLE-UDP firewall for:
IP: 144.76.140.250
Ports: 2302:2303

Below is a tcpdump of the communication generated by the given URL: 
http://www.wargamer-project.de/gameq/examples/index.php
14:26:48.483669 IP 69.5.11.3.45984 > 144.76.140.250.2302: UDP, length 9
14:26:48.484401 IP 69.5.11.3.53824 > 144.76.140.250.2303: UDP, length 9
14:26:48.485016 IP 69.5.11.3.57258 > 144.76.140.250.2302: UDP, length 9
14:26:48.641790 IP 144.76.140.250.2303 > 69.5.11.3.53824: UDP, length 9
14:26:49.177411 IP 69.5.11.3.48164 > 144.76.140.250.2302: UDP, length 25
14:26:49.178047 IP 69.5.11.3.38677 > 144.76.140.250.2302: UDP, length 7
14:26:49.178782 IP 69.5.11.3.57616 > 144.76.140.250.2302: UDP, length 7
14:26:49.179551 IP 69.5.11.3.51362 > 144.76.140.250.2303: UDP, length 25
14:26:49.180534 IP 69.5.11.3.60543 > 144.76.140.250.2303: UDP, length 9
14:26:49.180966 IP 69.5.11.3.49859 > 144.76.140.250.2303: UDP, length 9
14:26:49.181733 IP 69.5.11.3.53435 > 144.76.140.250.2302: UDP, length 25
14:26:49.182428 IP 69.5.11.3.52220 > 144.76.140.250.2302: UDP, length 7
14:26:49.183070 IP 69.5.11.3.33586 > 144.76.140.250.2302: UDP, length 7
14:26:49.324466 IP 144.76.140.250.2303 > 69.5.11.3.51362: UDP, length 166
14:26:49.324532 IP 144.76.140.250.2303 > 69.5.11.3.60543: UDP, length 135
14:26:49.324806 IP 144.76.140.250.2303 > 69.5.11.3.49859: UDP, length 649

As you can see, communication on port 2303 is now flowing in both directions,
however the remote server is not properly responding back from port 2302... I have
confirmed this at our Level3 edge router where the packets are coming from for the
replies...

At this point you will need to check out why 144.76.140.250:2302 is not responding
back to your queries...
```

Despite the "improper response" on 2302, the actual query works fine, and all the proper data is received.

Thanks for your assistance, Austin!

**Comment by Austinb** [comment]
Thank you for the update.  It is not too surprising that the protocols are not perfect when it comes to the packets being sent out.  If the person helping you has any input on how to make the packets more complete please post them here.
